akka-actor/src/main/scala/akka/util/ByteString.scala

```diff
@@ -364,21 +364,57 @@ object ByteString {
   }
 
   /**
+   * This class helps to optimise the access time for linear access such as looping all items of [[ByteStrings]].
+   *
+   * @param lastIndex the index of the last access
+   * @param arrayIndex the array index of the last access
+   * @param seen the amount of elements being seen, excluding the elements in current array
+   */
+  private final case class LinearScanHint(lastIndex: Int, arrayIndex: Int, seen: Int)
```

Comment:
👍 
However, this solution have two defects: 
* need more extra space.
* have to loop from starting point when the index moves to the next array.